### DISCUSSION TITLE
C# programming language.

### DISCUSSION THREAD

**Discussion by 111ukasz111** [o3de/o3de#2044] [Ideas]
**Describe the solution you'd like**
Many Unity users will migrate to o3de.

**Describe alternatives you've considered**
-

**Additional context**
C# is a very popular language.

**Comment by Darkinggq** [comment]
why do you need `C#` but o3de is a programming language in `Lua` similar to Python syntax

**Reply by HydrogenC** [reply]
C# features a greatly improved runtime speed and rich third party package support compared with lua.

**Comment by S3r1ous** [comment]
i am no expert in these things but anything that avoids complexity blowing up as things get added/improved is probably good i think
switching languages/major components etc is big thing, its not as easy as it looks and often whatever perceived benefits either aren't worth it or get outweighed by whats already there, you need really good reasons to do these big changes and crystal clear or as close to it understanding why
Lua, C# languages like that, have a lot of uses and in a lot of things benefit the workflow, but if i could i would prefer do things with just a few very specific tools to avoid all the compatibility etc headache
just my 2 cents

**Comment by olegbc1** [comment]
Should probably leave it to a SIGs but FWIW, transitions from C# to C libs will cause significant performance degradation. C# makes a lot of sense in Unity since so much of it is written in C# and execution can just stay in CIL, but for an engine written in C++, implementing C# for scripting just wouldn't be optimal. There are things other than performance to consider of course...

**Reply by harryisgamer** [reply]
Unity is actually written in both C++ and C#, with core components written in C++ and stuff like the editor and terrain system written in C#. Using C++ for engine code and C# for game code works quite nicely, since C++ is far better for writing engines than C# is, but C# is nicer to write game code in (IMHO). I am definitely interested in official C# support in O3DE. Lua works well as a scripting language, but being able to write an entire game in C# would be nice. But yes, performance wouldn't be as good as in C++.

**Reply by JeroMiya** [reply]
.net managed-unmanaged interop, as well as raw script performance, can be significantly faster than Lua and Lua interop, if implemented correctly and when using a modern dotnet runtime.

**Comment by ghost** [comment]
No, c# is trash

**Comment by thecrazy** [comment]
Here are a few valid questions:

1. Can Lua use a burst compiler?
2. Can Lua use IL2CPP to improve performance to c++ levels?
3. Does Lua have a HUGE user base of indy game developpers that learned the language over many years, ready to switch to a free alternative?
4. Does learning Lua give you any advantage in the job market?
5. Do you care about performance? [Lua vs C# Benchmark](https://programming-language-benchmarks.vercel.app/lua-vs-csharp)

You may want to revise your definition of trash.

**Reply by Qubrick** [reply]
> Here are a few valid questions:
> 
> 1. Can Lua use a burst compiler?
> 2. Can Lua use IL2CPP to improve performance to c++ levels?
> 3. Does Lua have a HUGE user base of indy game developpers that learned the language over many years, ready to switch to a free alternative?
> 4. Does learning Lua give you any advantage in the job market?
> 5. Do you care about performance? [Lua vs C# Benchmark](https://programming-language-benchmarks.vercel.app/lua-vs-csharp)
> 
> You may want to revise your definition of trash.

Both Burst Compiler and iL2CPP are parts of Unity technology. Other C# engine cannot just use them. I think even Microsoft didn't know how it was implemented. It's their secret sauce to bridge gap of AOT and increase performance for C# in Unity.

Lua has been used in Roblox. That game alone has huge amount of indie beginner developers. Lua also use in Core. Either are platforms with huge user base as well as developers.  I also found that many open source engines have integrated Lua in their engine. (Wicked Engine, The Forge, Pico-8, LÖVE, etc.)

**Comment by JeroMiya** [comment]
This would be a welcome addition to the engine, though I would suggest that rather than implementing C#/.NET support directly in the engine, it should instead be implemented as part of a broader script extensibility system allowing any runtime/language to create a plugin that opens up scripting in that language to the engine, including dotnet. One benefit of this would be that developers could use F# for instance, instead of C#, if they preferred. And it would open up the engine to other languages as well, e.g. visual scripting tools, python, JVM languages, rust, etc...

The engine should provide extension points that allow the custom scripting system to manage how scripts are identified (some languages are file based, but the dotnet adaptor should use full type names with assembly to allow multi-assembly projects and nuget support), how the engine queries for available scripts, how to serialize components and properties from scripts, and so on, etc... so that each scripting adaptor can adopt idiomatic best practices for that runtime or language.

Also, if the engine doesn't already do this, it should provide a machine readable set of metadata for the full FFI interface exposed to script engine adaptors, so that bindings can be autogenerated (similar to Silk.Net bindings, for example), reducing maintenance cost of such bindings.

Further, to make this work in a cross-platform way, the engine should support being loaded in as a library rather than the main entry point, as many runtimes will require that they be the entry point (particularly true of mobile/console/WebAssembly target platforms). O3DE may already work this way, I don't know.

**Reply by GunpowderGuy** [reply]
This engine being based around component system and not using inheritance much ( or in a simple way ) in the user facing API ( or so i gather from reading the documentation ) . Could mean that it could be particularly well suited for being used from other languages. 

To elaborate. Some languages like rust or haskell dont support inheritance, but it still is practical to write bindings to code from libraries that do if said code uses inheritance in a limited ammount of ways ( eg : mainly inheriting from a small number of classes ) .

Component systems are widely used across programming language to manage memory structures in a way easy for their memory management to cope with. Eg: rust uses them because it jells well with the borrow checker. F# and haskell uses them to hide mutation behind a functional api ( eg : https://github.com/bryanedds/Nu )

**Comment by JailbreakPapa** [comment]
Please see: https://github.com/WatchDogStudios/O3DESharp/tree/development
